src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java

```diff
@@ -451,6 +452,46 @@ void createDicomStoreInDataset(DicomPath dicomPath) throws DicomFuseException {
     }
   }
 
+  void renameDicomStoreInDataset(DicomPath oldDicomPath, DicomPath newDicomPath) {
+    Runnable renameRunnable = () -> {
+      if (newDicomPath.getDicomPathLevel() == DICOM_STORE) {
+        try {
+          if (isDicomStoreEmpty(oldDicomPath)) {
+            FuseDaoHelper
+                .deleteDicomStore(parameters.getFuseDAO(), parameters.getCloudConf(), oldDicomPath);
+            LOGGER.info("DICOM Store was deleted - " + oldDicomPath);
+            cache.getCachedDicomStores().remove(oldDicomPath.getDicomStoreId());
+            FuseDaoHelper
+                .createDicomStore(parameters.getFuseDAO(), parameters.getCloudConf(), newDicomPath);
+            String dicomStoreId = newDicomPath.getDicomStoreId();
+            DicomStore dicomStore = new DicomStore();
+            dicomStore.setDicomStoreId(dicomStoreId);
+            CachedDicomStore newCachedDicomStore = new CachedDicomStore(dicomStore);
+            cache.getCachedDicomStores().put(dicomStoreId, newCachedDicomStore);
+            LOGGER.info("DICOM Store was created - " + newDicomPath);
+          } else {
+            LOGGER.error("DICOM Store " + oldDicomPath + " should be empty");
+          }
+        } catch (DicomFuseException e) {
+          LOGGER.error("Rename error", e);
+        }
+      } else {
+        LOGGER.error("You can only rename DICOM Store folder");
+      }
+    };
+    Thread thread = new Thread(renameRunnable);
+    thread.start();
+  }
+
+  boolean isDicomStoreEmpty(DicomPath dicomPath) throws DicomFuseException {
+    Path oldPath = parameters.getMountPath().resolve(dicomPath.getDicomStoreId());
+    try (DirectoryStream<Path> stream = Files.newDirectoryStream(oldPath)) {
```

Comment:
Can we just send a QIDO request directly here instead of relying on fuse?

Reply:
I changed this.